WordPress/gutenberg
# Edit Post: Make sidebar header focusable for button focus normalization

Fixes #20759

This pull request seeks to resolve an issue where clicking the "Close Sidebar" in Firefox for macOS has two problems:

- Focus will not be returned to the corresponding header toolbar toggle button
- The "Open document settings" panel at the bottom of the sidebar remains visible

There are a number of contributing factors here, all of which ultimately stem from a [particular browser quirk](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus) in how Firefox for macOS treats clicks on button elements. Essentially, the problem is that clicking the sidebar dismiss button never registers as the sidebar ever having received focus, so it does not know to either return focus to the header, nor to dismiss the bottom panel of the sidebar.

In other browsers, the normal expected behavior is:

1. A click on the button bubbles as a focus event captured on `withFocusReturn` state:

https://github.com/WordPress/gutenberg/blob/d4f159a3da0738aa045e9969dba2c6c13a7f6a8d/packages/components/src/higher-order/with-focus-return/index.js#L105-L108

2. When the (inner) sidebar component is unmounted, it runs the focus return logic if focus was within. This is where Firefox fails, because at this point it does not consider focus as being within the element.

https://github.com/WordPress/gutenberg/blob/d4f159a3da0738aa045e9969dba2c6c13a7f6a8d/packages/components/src/higher-order/with-focus-return/index.js#L73-L84

3. The implementation of the sidebar explicitly shifts focus to the header toolbar toggle button, having the dual effect of shifting focus to a meaningful fallback, and hiding the sidebar panel.

https://github.com/WordPress/gutenberg/blob/d4f159a3da0738aa045e9969dba2c6c13a7f6a8d/packages/edit-post/src/components/sidebar/index.js#L33-L41

There is some prior art to this proposed solution, specifically in how the sibling inserter captures focus via click on the inserter button:

https://github.com/WordPress/gutenberg/blob/d4f159a3da0738aa045e9969dba2c6c13a7f6a8d/packages/block-editor/src/components/block-list/insertion-point.js#L166-L172

**Testing Instructions:**

Repeat Steps to Reproduce from #20759, verifying that the panel at the bottom of the sidebar no longer remains visible, and that focus is shifted to the sidebar toggle button in the header toolbar.
